### ISSUE TITLE
Retained entities not getting released

### ISSUE THREAD

**Issue by hahasss95** [sschmid/Entitas-CSharp#33]
Hi, I am quite confuse now. Is retained entities bad? Is there anyway to clear the retained entities or move it to reusable entities?

**Comment by sschmid** [comment]
Automatic Entity Reference Counting (AERC) was introduced to fix an internal bug (#25). It works automatically and ist managed by Entitas. Usually you don't have contact with this feature and you should never manually call entity.Retain() or entity.Release(). The ONLY exception to call these methods manually is when you want to keep a reference to an entity on your own, which is not recommended.
Basically you can happily forget, that this feature even exist ;)

**Comment by hahasss95** [comment]
Thank you for your explanation. At first, I didn't know that the order of creating the system will affect how the loop goes. So I do not need to worry about having a lot of retained entities?

**Comment by JuDelCo** [comment]
Try disabling systems at runtime to see which one is retaining the entities. Anyway... how much is a lot ?

**Comment by sschmid** [comment]
@hahasss95 I don't really understand what you mean. Could you explain what your concerns are? Again, consider AERC as an internal feature, you don't really have to care. It's all managed automatically for you.

**Comment by hahasss95** [comment]
I managed to have no retained entities as before that there were as much as around 50. I was afraid that it will slow down my game as there were no reusable entities as all the destroyed entities are being retained.

**Comment by sschmid** [comment]
Sorry, I have the feeling I still don't understand what the question is.

> ...there were as much as around 50

did you debug your game, or how do you know the retained entities count? Because there is no public access to this information.

> ... as all the destroyed entities are being retained.

pool.DestroyEntity() will release entities and therefore make them reusable.

To answer your initial question

> Is retained entities bad? Is there anyway to clear the retained entities or move it to reusable entities?

It's all managed for you automatically, you don't have to do anything. You don't even have to know that entities are being reused under the hood as this is just an performance and memory optimization. Call pool.CreateEntity() if you need one, call pool.DestroyEntity() if you want to destroy it ;) That's basically it. Anything else are implementation details.

> I was afraid that it will slow down my game...

AERC won't slow your game down ;)

Hope that helps.

**Comment by hahasss95** [comment]
Yeah I did debug my game to see the number of retained entities. Okay thank you. I will just let how the entities work and not change things that I do not know. Thanks for all the explanations! :)

**Comment by sschmid** [comment]
I mean, if you experience some unexpected behaviour or you think you found a bug, let me know. The expected behaviour is:
When you create a new entity
- _reusableEntities.Count == 0
- _retainedEntities.Count == 0

When you destroy this entity
- _reusableEntities.Count == 1
- _retainedEntities.Count == 0

Depending on your reactive systems and group observers _retainedEntities.Count can be > 0, but will be == 0 by the end of the loop.

**Comment by sschmid** [comment]
I think I know what you mean now! :)
I could reproduce what you explained. The problem was that the GroupObserver could potentially retain entities multiple times. I fixed that in release [0.22.0](https://github.com/sschmid/Entitas-CSharp/releases/tag/0.22.2)

Let me know if it fixes your issue as well.
Thanks :)

**Comment by hahasss95** [comment]
Yeah, It seems to work fine now. No more retained entities.
